bsp4j/src/main/java/ch/epfl/scala/bsp4j/Protocol.xtend

```diff
@@ -530,6 +531,36 @@ class RunResult {
   }
 }
 
+@JsonRpcData
+class ExecutionEnvironmentParams {
+  @NonNull List<BuildTargetIdentifier> targets
+  new(@NonNull List<BuildTargetIdentifier> targets) {
+    this.targets = targets
+  }
+}
+
+@JsonRpcData
+class ClasspathItem {
+  @NonNull BuildTargetIdentifier target
+  @NonNull List<String> classpath
+  new(@NonNull BuildTargetIdentifier target, @NonNull List<String> classpath) {
+    this.target = target
+    this.classpath = classpath
+  }
+}
+
+@JsonRpcData
+class ExecutionEnvironmentResult {
+  @NonNull String workingDirectory
+  @NonNull Map<String, String> environmentVariables
```

Comment:
Can we include Java options here? `List<String> javaOptions`, which include system properties like `-D...` and memory options like `-X...`

Reply:
Done

Reply:
The current encoding doesn't allow unique working directories or environment variables per targets. Would it make sense to move those into `JvmEnvironmentEntry`? At the risk of shaving yaks, I think the name should end with "Item" instead of "Entry" and organized like this

```scala
case class JvmEnvironmentItem(workingDirectory: Path, environmentVariables: Map[String, String], classpath, ...)
class JvmEnvironmentResult(items: List[JvmEnvironmentItem])
```

I agree we should use "entry" for single elements of a classpath.

Reply:
This comes from my understanding of how it is done in Bloop. Working directory and environment variables are stored in State class, so if your endpoint looks like this:
```
    ifInitialized(None) { (state: State, logger: BspServerLogger) =>
      mapToProjects(params.targets, state) match {
```
the working directory is stored in `state.commonOptions.env` and environment variables are in `state.commonOptions.workingDirectory`. 

I don't know if it makes sense to put it into `Entry` class,  but it could. If you think it'll be better,  in current bloop implementation I'd just copy the same value to all entries in current bloop implementation. Pleas just let me know what's your view on this.

Reply:
In Bloop, each `Project` can have a unique working directory https://github.com/scalacenter/bloop/blob/dfd03f5333c2fc292db78bfcd4205a0cdd828967/frontend/src/main/scala/bloop/data/Project.scala#L66

It's the same situation in sbt, where `baseDirectory.in(run).value` is used for running in forked JVMs.

As for environment variables, Bloop currently only uses the environment variables that are passed from the terminal interface. However, for consistency with other configuration I don't see a reason why we should not allow per-target variables.

Reply:
I agree with @olafurpg's suggestion that working directory and env variables could make more sense as part of the `JvmEnvironmentEntry` since they can be project-specific.

Regarding your comments of how we do it in bloop:
* We don't use `state.commonOptions.workingDirectory` as the working directory of the project we run but the `directory` configuration field that is configured at the project lefvel. 
* You're right we currently allow the client to pass the environment variables from the place the CLI is run via `state.commonOptions.env`, but in the future we will enable build tools to export environment variables to the configuration file on a per-project basis. This is a common idiom so we want to support it.

Reply:
Ok, thanks, I'm just changing this